server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -481,29 +481,42 @@ private Builder initializeEmpty(IndexMetadata indexMetadata, UnassignedInfo unas
                         )
                     );
                 }
-                shards.put(shardNumber, indexShardRoutingBuilder.build());
+                shards[shardNumber] = indexShardRoutingBuilder;
             }
             return this;
         }
 
         public Builder addReplica() {
-            for (var shardNumber : shards.keySet()) {
-                ShardId shardId = new ShardId(index, shardNumber);
+            if (shards == null) {
+                return this;
+            }
+            for (IndexShardRoutingTable.Builder existing : shards) {
+                if (existing == null) {
+                    continue;
+                }
                 // version 0, will get updated when reroute will happen
-                ShardRouting shard = ShardRouting.newUnassigned(
-                    shardId,
-                    false,
-                    PeerRecoverySource.INSTANCE,
-                    new UnassignedInfo(UnassignedInfo.Reason.REPLICA_ADDED, null)
+                existing.addShard(
+                    ShardRouting.newUnassigned(
+                        existing.shardId(),
+                        false,
+                        PeerRecoverySource.INSTANCE,
+                        new UnassignedInfo(UnassignedInfo.Reason.REPLICA_ADDED, null)
+                    )
                 );
-                shards.put(shardNumber, new IndexShardRoutingTable.Builder(shards.get(shard.id())).addShard(shard).build());
             }
             return this;
         }
 
         public Builder removeReplica() {
-            for (var shardId : shards.keySet()) {
-                IndexShardRoutingTable indexShard = shards.get(shardId);
+            if (shards == null) {
```

Comment:
Like for `addReplica`, I think we always have this filled in here and can instead `assert shards != null`?

Reply:
++ done